contracts/wallet/WalletFactory.sol

```diff
@@ -56,38 +69,135 @@ contract WalletFactory is Owned, Managed {
-            // add the factory to the modules so it can claim the reverse ENS
-            address[] memory extendedModules = new address[](_modules.length + 1);
-            extendedModules[0] = address(this);
-            for(uint i = 0; i < _modules.length; i++) {
-                extendedModules[i + 1] = _modules[i];
-            }
-            // initialise the wallet with the owner and the extended modules
-            BaseWallet(wallet).init(_owner, extendedModules);
-            // register ENS
-            registerWalletENS(wallet, _label);
-            // remove the factory from the authorised modules
-            BaseWallet(wallet).authoriseModule(address(this), false);
-        } else {
-            // initialise the wallet with the owner and the modules
-            BaseWallet(wallet).init(_owner, _modules);
+        configureWallet(BaseWallet(wallet), _owner, _modules, _label, address(0));
+    }
+
+    /**
+     * @dev Lets the manager create a wallet for an owner account.
+     * The wallet is initialised with a list of modules, a first guardian, and an ENS..
+     * The wallet is created using the CREATE opcode.
+     * @param _owner The account address.
+     * @param _modules The list of modules.
+     * @param _label ENS label of the new wallet, e.g. franck.
+     * @param _guardian The guardian address.
+     */
+    function createWalletWithGuardian(
+        address _owner,
+        address[] calldata _modules,
+        string calldata _label,
+        address _guardian
+    )
+        external
+        onlyManager
+        guardianStorageDefined
+    {
+        validateInputs(_owner, _modules, _label);
+        Proxy proxy = new Proxy(walletImplementation);
+        address payable wallet = address(proxy);
+        configureWallet(BaseWallet(wallet), _owner, _modules, _label, _guardian);
+    }
+
+    /**
+     * @dev Lets the manager create a wallet for an owner account at a specific address.
+     * The wallet is initialised with a list of modules and an ENS.
+     * The wallet is created using the CREATE2 opcode.
+     * @param _owner The account address.
+     * @param _modules The list of modules.
+     * @param _label ENS label of the new wallet, e.g. franck.
+     * @param _salt The salt.
+     */
+    function createCounterfactualWallet(
+        address _owner,
+        address[] calldata _modules,
+        string calldata _label,
+        bytes32 _salt
+    )
+        external
+        onlyManager
+    {
+        validateInputs(_owner, _modules, _label);
+        bytes32 newsalt = keccak256(abi.encodePacked(_salt, _owner, _modules));
+        bytes memory code = abi.encodePacked(type(Proxy).creationCode, uint256(walletImplementation));
+        address payable wallet;
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            wallet := create2(0, add(code, 0x20), mload(code), newsalt)
+            if iszero(extcodesize(wallet)) { revert(0, returndatasize) }
```

Comment:
Instead of checking the code at address, can we simply have a `require(wallet != address(0), "error message"`. `create2` returns an address when successful and `0` if not.

Reply:
This is what I used initially but playing with CREATE2 in another project I noticed that create2 sometimes returns an address when no code is deployed. I did not investigate further but decided to use the check used by openzeppelin and other projects.

Reply:
Open Zeppelin don't use that though https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Create2.sol#L26 ?

Reply:
I guess they no longer do https://blog.openzeppelin.com/getting-the-most-out-of-create2/.

But it doesn't change the fact that I witnessed some unreliable behavior when checking solely on the returned address so I'll rather keep checking on the bytecode.